design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,438 @@
+    purpose is to guarantee pod restart any time the key is rotated.
+8. IFF keys were created/rotated, Rook updates the appropriate cephx key status(es)
+    1. `status...keyGeneration` is taken to be `spec...keyGeneration`
+    2. `status...keyCephVersion` is taken to be `cephImageVersion`
+
+### CephCluster example
+
+1. CephCluster daemons are all local, so the reconcile reads
+    `spec.security.cephx.local` to determine if the reconcile should rotate.
+2. When key rotation is indicated:
+    1. Rotate the admin key. This is the first key created by Rook and will be the first rotated.
+        This should be as easy as rotating it and updating the local reference. However, depending
+        on how child reconciliations are updated with the admin key, it may be necessary to restart
+        the Rook operator reconcile manager (all reconcile routines) after this rotation.
+    2. <!-- TBD --> Update the CephCluster status with the admin key info (with internal retry)
+    3. Rotate the mon key. Mon key rotation requires a special process:
+        1. Update mon deployments first without rotating the key
+        2. Rotate the key
+        3. Restart the mon deployments to cause them to get the rotated key
+    4. <!-- TBD --> Update the CephCluster status with the mon key info (with internal retry)
+    5. For each mgr, rotate the key and update/restart the mgr daemon
+    6. <!-- TBD --> Update the CephCluster status with the mgr key info (with internal retry)
+    7. For each OSD, rotate the key
+        1. OSDs don't have an associated keyring secret, so we must make sure all OSD deployments'
+            "activate" init container gets the latest key and updates it for the "osd" pod
+    8. <!-- TBD --> Update the CephCluster status with the OSD key info (with internal retry)
+
+With separated daemon key info tracking, the status will look like so:
+
+```yaml
+status:
+  # ...
+  cephx:
+    admin:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mon:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mgr:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    osd:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    csi: {} # discussed more below, but some CSI keys and need to be in this status
+```
+
+<!-- TODO: DISCUSS -->
+
+I believe we can track all CephCluster key rotations as one and rely on operator logs and error
+events for debugging, but we should still separate out CSI rotations since those can be initiated
+separately.
+
+```yaml
+status:
+  # ...
+  cephx:
+    local:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    csi: {} # discussed more below, but some CSI keys and need to be in this status
+```
+
+### CSI example
+
+This is complicated, and I need some input. My understanding is that CSI keys are are always updated
+as part of a CephCluster, CephRBDMirror, or CephFilesystemMirror reconcile.
```

Comment:
CSI keys are always updated as prt of CephCluster Reconcile. i think its nothing to do with CephRBDMirror nor CephFilesystemMirror.